pkg/registry/service/allocator/controller/repair.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service/allocator"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// Repair is a controller loop that periodically examines all service PortalIP allocations
+// and logs any errors.
+// TODO: allocate new IPs if necessary
+// TODO: perform repair?
+type Repair struct {
+	interval time.Duration
+	registry service.Registry
+	network  *net.IPNet
+	alloc    service.IPRegistry
+}
+
+// NewRepair creates a controller that periodically ensures that all portalIPs are uniquely allocated across the cluster
+// and generates informational warnings for a cluster that is not in sync.
+func NewRepair(interval time.Duration, registry service.Registry, network *net.IPNet, alloc service.IPRegistry) *Repair {
+	return &Repair{
+		interval: interval,
+		registry: registry,
+		network:  network,
+		alloc:    alloc,
+	}
+}
+
+// RunUntil starts the controller until the provided ch is closed.
+func (c *Repair) RunUntil(ch chan struct{}) {
+	util.Until(func() {
+		if err := c.RunOnce(); err != nil {
+			util.HandleError(err)
+		}
+	}, c.interval, ch)
+}
+
+// RunOnce verifies the state of the portal IP allocations and returns an error if an unrecoverable problem occurs.
+func (c *Repair) RunOnce() error {
+	latest, err := c.alloc.Get()
+	if err != nil {
+		return fmt.Errorf("unable to refresh the service IP block: %v", err)
+	}
+
+	ctx := api.WithNamespace(api.NewDefaultContext(), api.NamespaceAll)
+	list, err := c.registry.ListServices(ctx)
```

Comment:
> On May 4, 2015, at 8:03 PM, Alex Robinson <notifications@github.com> wrote:
> 
> In pkg/registry/service/allocator/controller/repair.go:
> 
> > +	util.Until(func() {
> > +		if err := c.RunOnce(); err != nil {
> > +			util.HandleError(err)
> > +		}
> > +	}, c.interval, ch)
> > +}
> > +
> > +// RunOnce verifies the state of the portal IP allocations and returns an error if an unrecoverable problem occurs.
> > +func (c *Repair) RunOnce() error {
> > +	latest, err := c.alloc.Get()
> > +	if err != nil {
> > +		return fmt.Errorf("unable to refresh the service IP block: %v", err)
> > +	}
> > +
> > +	ctx := api.WithNamespace(api.NewDefaultContext(), api.NamespaceAll)
> > +	list, err := c.registry.ListServices(ctx)
> For my own learning - is it typically better to do a full list against the registry rather than a watch with some sort of cache (like pkg/client/cache.Store)?
> 
For this controller we get no benefit from incremental calculation (need to calculate the full set to detect deletes), so the cache store is a small percentage more expensive than the list (under the covers store also does periodic watches).  We probably want to abstract a simple "every X seconds calculate something based on the full set" controller for when the set starts to grow, but for now cache.store doesn't make this easier to read or implement and is more expensive.
> —
> Reply to this email directly or view it on GitHub.
>